### PR TITLE
Apr 2733 enable ukprn address reapplication

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.ApplyService.Data/Queries/OversightReviewQueries.cs [hunk 1]
```diff
@@ -237,7 +237,7 @@ public async Task<ApplicationOversightDetails> GetOversightApplicationDetails(Gu
                               FROM Apply apply
 	                      INNER JOIN Organisations org ON org.Id = apply.OrganisationId
                           LEFT JOIN OversightReview outcome ON outcome.ApplicationId = apply.ApplicationId
-						  LEFT OUTER JOIN contacts on contacts.ApplyOrganisationId = org.Id
+						  LEFT OUTER JOIN contacts on contacts.ApplyOrganisationId = org.Id and apply.CreatedBy = contacts.id
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Data/Queries/OversightReviewQueries.cs:240]: I'm not sure what this is doing but looks all part of the process

File: src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            _applicationApiClient = applicationApiClient;
+        }
+
+        public async Task<bool> ReapplicationAllowed(Guid signInId, Guid? organisationId)
+        {
+            if (organisationId == null)
+                return true;
+
+            var applications = await _applicationApiClient.GetApplications(signInId, true);
+            var application = applications.OrderByDescending(x => x.UpdatedAt)
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs:25]: Needs Null Coalesce to stop a potential Null Reference Exception
`applications?.OrderByDescending(x => x.UpdatedAt)`

File: src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            return application.ApplicationStatus == ApplicationStatus.Rejected &&
+                   applyDetails?.RequestToReapplyMade == true;
+        }
+
+        public async Task<string> ReapplicationUkprnForUser(Guid signInId)
+        {
+            var applications = await _applicationApiClient.GetApplications(signInId, true);
+            var applicationsRejectedAndRequestToReapplyMade = applications.FirstOrDefault(x =>
+                x.ApplicationStatus == ApplicationStatus.Rejected &&
+                x?.ApplyData?.ApplyDetails?.RequestToReapplyMade == true);
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs:40]: nit picking but x cannot but null in this lambda expression

File: src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+        public async Task<bool> ApplicationInFlightWithDifferentUser(Guid signInId, string ukprn)
+        {
+            var applications = await _applicationApiClient.GetApplicationsByUkprn(ukprn);
+
+            var contact = await _applicationApiClient.GetContactBySignInId(signInId);
+            var contactId = contact?.Id.ToString()?.ToUpper();
+
+            var applicationsPresentAgainstDifferentUser = applications.OrderByDescending(x => x.UpdatedAt)
+                .Where(x => x?.ApplyData?.ApplyDetails?.UKPRN == ukprn && x.CreatedBy!=contactId && x.ApplicationStatus!=ApplicationStatus.Cancelled);
```
Comment (martin-lush) [inline src/SFA.DAS.ApplyService.Web/Services/ReapplicationCheckService.cs:53]: and here aswell

**Review by martin-lush — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApplyService.Data/Queries/OversightReviewQueries.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Data/Queries/OversightReviewQueries.cs:240]: ah that's a left over from the different email> application stuff, I'm taking it out

**Review by martin-lush — APPROVED** [review]

**Review by martin-lush — APPROVED** [review]

**Pull request merged**